### PR TITLE
Support for plugin manifest

### PR DESCRIPTION
**Pull request by lemillermicrosoft**
### Motivation and Context
Developers do not always have access to the kernel code or access to change it. Plugin developers will need to be able to create their plugins and deploy those. After they do that, they will use the manifest file with the description in it to store into a centralized location. This location will need to be fast to query and retrieve plugins that will help solve the ask.

In support of #1728 

### Description
This pull request adds a new feature to the sequential planner that allows users to specify a custom function to get the available functions for planning. This can be useful for scenarios where the default function is not sufficient or where users want to use a different set of functions for different queries. The pull request also adds support for importing a plugin manifest from a URL, which is a JSON file that lists the names and URLs of chat GPT plugins. This can make it easier for users to discover and use multiple plugins without having to import them individually.

Details:
- Add a new property to SequentialPlannerConfig called GetAvailableFunctionsAsync, which is a delegate that takes a config and a semantic query and returns an ordered enumerable of FunctionView objects. If this property is null, the default function is used.
- Modify SKContextSequentialPlannerExtensions.GetManualAsync to use the custom function if provided, or the default function otherwise.
- Add a TODO comment in SequentialPlanner to note that the kernel needs to be able to load functions on-demand if a custom function is used.
- Add a new extension method to IKernel called ImportPluginManifestFromUrlAsync, which takes a URL of a plugin manifest and imports all the plugins listed in it. The plugin manifest is a JSON file that follows the PluginManifest model, which has a list of RemotePlugin objects, each with a name and a URL.
- Add a new model class called PluginManifest, which represents the plugin manifest JSON file, and a nested class called RemotePlugin, which represents a single plugin entry.
- Add a private helper method to KernelChatGptPluginExtensions called ParsePluginJson, which deserializes the plugin manifest JSON file and validates it.
### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [ ] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Skills/Skills.OpenAPI/Model/PluginManifest.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+/// <summary>
+/// A plugin manifest that lists plugins and their names.
+/// </summary>
+public sealed class PluginManifest
+{
+    public List<RemotePlugin> Plugins { get; } = new List<RemotePlugin>();
+}
+
+public sealed class RemotePlugin
```
Comment (SergeyMenshykh) [inline dotnet/src/Skills/Skills.OpenAPI/Model/PluginManifest.cs:16]: Please add XML comments to this class.

File: dotnet/src/Skills/Skills.OpenAPI/Extensions/KernelChatGptPluginExtensions.cs [hunk 2]
```diff
@@ -176,4 +206,12 @@ private static string ParseOpenApiUrl(string gptPluginJson)
 
         return openApiUrl!;
     }
+
+    private static PluginManifest ParsePluginJson(string pluginJson)
+    {
+        var manifest = JsonSerializer.Deserialize<PluginManifest>(pluginJson, new JsonSerializerOptions { IncludeFields = true })
```
Comment (SergeyMenshykh) [inline dotnet/src/Skills/Skills.OpenAPI/Extensions/KernelChatGptPluginExtensions.cs:212]: The PluginManifest class/model does not declare any fields, so setting the flag IncludeFields may not be necessary.

File: dotnet/src/Skills/Skills.OpenAPI/Extensions/KernelChatGptPluginExtensions.cs [hunk 3]
```diff
@@ -24,6 +26,34 @@ namespace Microsoft.SemanticKernel;
 /// </summary>
 public static class KernelChatGptPluginExtensions
 {
+    public static async Task<IDictionary<string, ISKFunction>?> ImportPluginManifestFromUrlAsync(
```
Comment (SergeyMenshykh) [inline dotnet/src/Skills/Skills.OpenAPI/Extensions/KernelChatGptPluginExtensions.cs:29]: XML comment is missing.

Comment (SergeyMenshykh) [inline dotnet/src/Skills/Skills.OpenAPI/Extensions/KernelChatGptPluginExtensions.cs:29]: This method reads/parses a custom plugin registry JSON document that is not part of the ChatGPT/OpenAI plugin specification, as far as I know. Therefore, I propose moving it out from the KernelChatGptPluginExtensions class to another class, such as Kernel**Custom**PluginExtensions or any other class that will group similar functionality together and provide better context.

Comment (lemillermicrosoft) [reply dotnet/src/Skills/Skills.OpenAPI/Extensions/KernelChatGptPluginExtensions.cs:29]: Yeah agreed! I whipped this up quickly to start getting just this kind of feedback. Thanks!

File: dotnet/src/Extensions/Planning.SequentialPlanner/SequentialPlanner.cs [hunk 4]
```diff
@@ -72,6 +72,8 @@ public async Task<Plan> CreatePlanAsync(string goal)
 
         try
         {
+            // TODO - If a custom function is used to determine which functions to use, the kernel needs to be able to load them on-demand somehow.
```
Comment (lemillermicrosoft) [inline dotnet/src/Extensions/Planning.SequentialPlanner/SequentialPlanner.cs:75]: TODO for this PR/design.

Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.SequentialPlanner/SequentialPlanner.cs:75]: Adding another function override for loading functions for now. Will see how that looks/feels.

File: dotnet/src/Extensions/Planning.SequentialPlanner/skprompt.txt [hunk 5]
```diff
@@ -2,112 +2,22 @@ Create an XML plan step by step, to satisfy the goal given.
...
+2. Before using any function in a plan, check that it is present in the [AVAILABLE FUNCTIONS] list. If it is not, do not use it.
 3. Only use functions that are required for the given goal.
-4. A function has a single 'input' and a single 'output' which are both strings and not objects.
-5. The 'output' from each function is automatically passed as 'input' to the subsequent <function>.
-6. 'input' does not need to be specified if it consumes the 'output' of the previous function.
-7. To save an 'output' from a <function>, to pass into a future <function>, use <function.{FunctionName} ... setContextVariable: "<UNIQUE_VARIABLE_KEY>"/>
-8. To save an 'output' from a <function>, to return as part of a plan result, use <function.{FunctionName} ... appendToResult: "RESULT__<UNIQUE_RESULT_KEY>"/>
-9. Append an "END" XML comment at the end of the plan.
+4. A function has at least a single 'input' and a single 'output' which are both strings and not objects. Additional parameters may be required for some functions as defined in the [AVAILABLE FUNCTIONS] list.
```
[the remark this answers is not in the document]
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.SequentialPlanner/skprompt.txt:7]: Pulled out the sequential specific changes into #1727 -- will reduce this PR to be focused on PlanManifest *proposal*/draft.

**Comment by lemillermicrosoft** [comment]
Thanks for your PR and contribution. For now, we are closing old drafts. Please re-open as appropriate.

**Pull request closed without merging**